### ISSUE TITLE
Flaky foreachPar_ Test

### ISSUE THREAD

**Issue by adamgfraser** [zio/zio#2642]
@horned-sphere Can you take a look?

```
[info] ZIO Test
[info] - ZIOSpec
[info]   - foreachPar_
[info]     - accumulates errors
[info]       Set(1, 2) did not satisfy equalTo(Set(1, 2, 3))
```

https://circleci.com/gh/zio/zio/50708?utm_campaign=workflow-failed&utm_medium=email&utm_source=notification

**Comment by horned-sphere** [comment]
@adamgfraser No problem. I'll take a look after work today.

**Comment by adamgfraser** [comment]
@horned-sphere Thanks!

**Comment by horned-sphere** [comment]
@adamgfraser Unfortunately, this was fairly unpleasant. foreachPar_ currently starts stopping after it hits the first error and so doesn't wait for any other tasks that have failed to register their results. That means that the selection of errors that get accumulated, even for uninterruptible tasks that have already started, is essentially random.

I've manged to fix it by splitting apart the wait for completion and the wait for failure. Sadly, it takes away a lot of the elegance of the original.

I've also beefed up the test case so it waits for all the tasks to start before they begin failing. I ran it continuously for a minute and it didn't fail so I think it's fairly watertight now. 

A PR is on its way.
